Tests/ParsingTests/ManyTests.swift

```diff
@@ -89,4 +89,21 @@ class ManyTests: XCTestCase {
     )
     XCTAssertEqual(Substring(input), "")
   }
+
+  func testFailureIfNoProgress() {
+     var input = " Hello, World!"[...]
+     XCTAssertEqual(
+       Many(CharacterSet.alphanumerics).parse(&input),
+       []
```

Comment:
@iampatbrown I think the question @stephencelis was asking about parsers that succeed without consuming was related to this point.

I started reworking #73 because my assumptions were wrong (I checked for `nil` instead of `[]`). 
In this case, `.alphanumerics` succeed in parsing the empty string `""`. Strictly speaking, the returned value should be `[""]`, without consuming the string. This is absolutely not what I would expect while using `Many`, but since `CharacterSet` cannot fail as a parser, and `Many` reports the successes of its upstream, I think this is what it should return. Returning `[]` implies that `.alphanumerics` succeeded 0 times, which is not the case.

In @iampatbrown example, I absolutely expect the ipv6 parser to return the empty components as empty strings, but in this test example, I would expect `[]`, as I would be trying to parse alpha-numeric chars and found a space to begin with.

This kind of ambiguity seems to happen whenever the upstream parser output is a collection and can succeed with an empty collection. This is furthermore problematic with `Many` itself when the empty collection results from parsing the input without consuming, which is what #73 and #74 are trying to handle. 

Maybe this should be formalized at a higher level. I initially commented on #72 because I was surprised that `CharacterSet` parser would never fail. I suspect many users would be in the same position. Maybe one can introduce a higher order `NonEmpty` parser to wrap a parser of collection into a failing parser when it parses an empty collection?. This behavior is achievable by mapping the results only a posteriori, but it would maybe be clearer for the user if it would happen at the parser type level, as the behavior of a `NonEmpty<CharacterSet>` parser feels unambiguous. @stephencelis, do you want me to experiment with this?

Reply:
Ah okay. I think I get what you're saying now. 
These examples are inconsistent
```swift
func testNonEmptyString() {
  let input = "123"[...]
  let upstream = CharacterSet.alphanumerics.parse(input)
  let many = Many(CharacterSet.alphanumerics).parse(input)
  XCTAssertEqual(upstream.output, many.output?.first)
} // Passed

func testEmptyString() {
  let input = ""[...]
  let upstream = CharacterSet.alphanumerics.parse(input)
  let many = Many(CharacterSet.alphanumerics).parse(input)
  XCTAssertEqual(upstream.output, many.output?.first) 
} // Failed
```

Reply:
I've proposed #77. Using `.nonEmpty` on `CharacterSet.alphanumerics`, one should expect:
```swift
Many(.alphanumerics).parse(" Hello, World!") == (output: [""], rest: " Hello, World!")
Many(.alphanumerics.nonEmpty).parse(" Hello, World!") == (output: [], rest: " Hello, World!")

Many(.alphanumerics).parse("123") == (output: ["123"], rest: "")
Many(.alphanumerics).parse("") == (output: [""], rest: "")
Many(.alphanumerics.nonEmpty).parse("") == (output: [], rest: "")

.alphanumerics.parse("123") == (output: "123", rest: "")
.alphanumerics.parse("") == (output: "", rest: "")
.alphanumerics.nonEmpty.parse("") == (output: nil, rest: "")

Many(.alphanumerics.nonEmpty).nonEmpty.parse("") == (output: nil, rest: "")
```

Reply:
@tgrapperon Definitely feel free to explore! We currently have things like `atLeast`/`atMost`/`minimum`/`maximum` options on `Many`/`Prefix`, but perhaps the concept could be generalized.